File: code_samples/section13/example_4_frequency_ranked/trie_autocomplete_ranked.js
// trie_autocomplete_ranked.js
//
// This Node.js script implements a trie (prefix tree) that supports frequency-ranked
// autocomplete. It loads words from two files:
// - ..\data\words.txt      (baseline dictionary words)
// - ..\data\frequency.txt  (additional occurrences / usage events)
// Each time a word is inserted, the terminal node's frequency counter is incremented.
//
// Ranking rules (as implemented):
// - Higher frequency first
// - If frequencies tie, lexicographically smaller word first
//
// Key constraints:
// - Only letters 'a'–'z' are accepted (case-insensitive).
// - If any character in a word is not a–z, insert() rejects the entire word.
// - If any character in the prefix is not a–z, autocomplete() returns [].
//
// Implementation approach:
// - DFS collects ALL completions under the prefix
// - Sort by frequency desc, then word asc
// - Slice to `limit` results
//
// Note:
// - Because this collects all completions before sorting, it can be memory/time heavy
//   for prefixes with many matches.

const fs = require("fs"); // File system API for reading dictionary/frequency files

/*
 * Node
 *
 * A trie node contains:
 * - children: array of 26 child references (a..z)
 * - freq: frequency count for the complete word ending at this node
 * - end: whether a complete word ends at this node
 *
 * Frequency is meaningful only when end === true.
 */
class Node {
  constructor() {
    this.children = Array(26).fill(null); // Child pointers (indices 0..25)
    this.freq = 0;                        // Word frequency counter (terminal nodes)
    this.end = false;                     // End-of-word marker
  }
}

/*
 * Trie
 *
 * Supports:
 * - insert(word): insert a word and bump its frequency at the terminal node
 * - autocomplete(prefix, limit): return top ranked completions for prefix
 */
class Trie {
  constructor() { this.root = new Node(); }

  /*
   * Convert a lowercase letter to child index [0..25].
   *
   * Parameters:
   * - c: single-character string
   *
   * Returns:
   * - 0..25 if valid 'a'..'z'
   * - -1 if invalid
   */
  idx(c) { return c >= "a" && c <= "z" ? c.charCodeAt(0) - 97 : -1; }

  /*
   * Insert a word into the trie and increment its terminal frequency.
   *
   * Behavior:
   * - Lowercases input (case-insensitive)
   * - Rejects the entire word if any char is not 'a'..'z'
   * - Creates missing nodes lazily
   * - Marks terminal node as end-of-word and increments freq
   *
   * Parameters:
   * - word: word to insert
   */
  insert(word) {
    let cur = this.root;

    // Iterate through lowercased characters
    for (const ch of word.toLowerCase()) {
      const i = this.idx(ch);

      // Reject entire word if invalid character encountered
      if (i < 0) return;

      // Allocate child if missing
      if (!cur.children[i]) cur.children[i] = new Node();

      // Advance to child node
      cur = cur.children[i];
    }

    // Mark complete word and bump frequency
    cur.end = true;
    cur.freq++;
  }

  /*
   * Autocomplete for a given prefix, ranked by frequency then word.
   *
   * Parameters:
   * - prefix: prefix to autocomplete
   * - limit: maximum number of suggestions to return
   *
   * Returns:
   * - Array of [word, freq] pairs, sorted best-first, truncated to `limit`
   *
   * Steps:
   * 1) Walk to the trie node for the prefix
   * 2) DFS the subtree collecting [word, freq] for terminal nodes
   * 3) Sort by frequency desc, then word asc
   * 4) Return first `limit` items
   *
   * Notes:
   * - This DFS does not short-circuit at `limit`; it collects everything, then sorts.
   */
  autocomplete(prefix, limit) {
    let cur = this.root;

    // Walk to the prefix node
    for (const ch of prefix.toLowerCase()) {
      const i = this.idx(ch);

      // Invalid prefix char or missing path => no suggestions
      if (i < 0 || !cur.children[i]) return [];

      cur = cur.children[i];
    }

    // Collected completions: [word, freq]
    const out = [];

    // Mutable character buffer seeded with lowercased prefix
    const buf = prefix.toLowerCase().split("");

    /*
     * DFS traversal collecting terminal words.
     *
     * Parameters:
     * - n: current node
     *
     * Behavior:
     * - If n.end is true, emit current buf as a candidate completion with n.freq
     * - Visit children in 'a'..'z' order (lexicographic traversal)
     * - Use buf push/pop to build words without allocating new strings each step
     */
    const dfs = (n) => {
      // If a complete word ends here, capture it
      if (n.end) out.push([buf.join(""), n.freq]);

      // Explore children in lexicographic order
      for (let i = 0; i < 26; i++) {
        if (n.children[i]) {
          // Append next letter
          buf.push(String.fromCharCode(97 + i));

          // Recurse into child subtree
          dfs(n.children[i]);

          // Backtrack: remove appended letter
          buf.pop();
        }
      }
    };

    // Traverse subtree beneath prefix
    dfs(cur);

    // Sort best-first:
    // - higher frequency first
    // - for ties, lexicographically smaller word first
    out.sort((a, b) => b[1] - a[1] || a[0].localeCompare(b[0]));

    // Return at most `limit` results (slice handles limit > out.length)
    return out.slice(0, limit);
  }
}

/* -------------------- Main script -------------------- */

// Build trie from dictionary + frequency inputs
const trie = new Trie();

/*
 * Load baseline dictionary words.
 *
 * Implementation:
 * - Read entire file as UTF-8 text
 * - Split by line endings (\n or \r\n)
 * - Insert each non-empty line
 *
 * Notes:
 * - Lines are not trimmed; a line with trailing spaces would include them and be rejected
 *   by insert() because spaces are not 'a'..'z'. (No changes requested—this is a note.)
 */
for (const w of fs.readFileSync("..\\data\\words.txt", "utf8").split(/\r?\n/))
  if (w) trie.insert(w);

/*
 * Load frequency/usage events.
 * Each non-empty line is treated as one occurrence of a word, bumping its freq.
 */
for (const w of fs.readFileSync("..\\data\\frequency.txt", "utf8").split(/\r?\n/))
  if (w) trie.insert(w);

/*
 * Query and print ranked autocomplete suggestions.
 *
 * Hard-coded example:
 * - prefix: "th"
 * - limit: 20
 *
 * Output format:
 * - word<TAB>(freq=N)
 */
for (const [w, f] of trie.autocomplete("th", 20))
  console.log(`${w}\t(freq=${f})`);

File: code_samples/section13/example_4_frequency_ranked/trie_autocomplete_ranked.java
package code_samples.section13.example_4_frequency_ranked;

import java.io.*;   // I/O primitives (BufferedReader, FileReader, IOException)
import java.util.*; // Collections (List, ArrayList, Map, Entry, etc.)

/*
 * trie_autocomplete_ranked
 *
 * This example implements a trie (prefix tree) that supports "frequency-ranked" autocomplete.
 * It loads words from two files:
 * - words.txt: a baseline dictionary
 * - frequency.txt: additional occurrences / usage events (each occurrence bumps frequency)
 *
 * Ranking rules (as implemented):
 * - Higher frequency first
 * - If frequencies tie, lexicographically smaller word first
 *
 * Key constraints:
 * - Only letters 'a'–'z' are accepted (case-insensitive).
 * - If any character in an inserted word is not a–z, insert() rejects that word.
 * - If any character in the prefix is not a–z, autocomplete() returns an empty list.
 *
 * Implementation approach:
 * - Collect ALL completions under the prefix using DFS
 * - Sort the collected list using the ranking rules
 * - Return the first `limit` items
 *
 * Note:
 * - This is simple but can be expensive for prefixes with many matches because it
 *   stores and sorts the entire subtree’s completions.
 */
public class trie_autocomplete_ranked {

    /*
     * Trie node.
     *
     * Fields:
     * - children: 26 child references (a–z)
     * - frequency: how many times this complete word has been inserted
     * - isEnd: whether a complete word ends at this node
     *
     * Frequency is only meaningful when isEnd == true.
     */
    private static class Node {
        Node[] children = new Node[26]; // Child pointers for letters a..z
        int frequency = 0;              // Word frequency counter (terminal nodes)
        boolean isEnd = false;          // End-of-word marker
    }

    // Root node of the trie (does not correspond to any character)
    private final Node root = new Node();

    /*
     * Convert a lowercase character to a trie child index.
     *
     * Parameters:
     * - c: expected to be 'a'..'z'
     *
     * Returns:
     * - 0..25 if valid
     * - -1 if invalid
     */
    private int idx(char c) {
        return (c >= 'a' && c <= 'z') ? c - 'a' : -1;
    }

    /*
     * Insert a word into the trie and bump its frequency.
     *
     * Behavior:
     * - Lowercases the word (case-insensitive)
     * - Rejects the entire word if any character is not 'a'..'z'
     * - Creates missing nodes lazily
     * - Marks terminal node as end-of-word and increments frequency
     *
     * Parameters:
     * - word: word to insert
     *
     * Frequency meaning:
     * - Each call to insert(word) increments that word’s frequency by 1.
     * - Loading multiple files means the same word can accumulate frequency across inputs.
     */
    public void insert(String word) {
        Node cur = root;

        // Normalize to lowercase and traverse each character
        for (char ch : word.toLowerCase().toCharArray()) {
            int i = idx(ch);

            // Reject word if invalid character encountered
            if (i < 0) return;

            // Allocate child node if missing
            if (cur.children[i] == null) cur.children[i] = new Node();

            // Advance in trie
            cur = cur.children[i];
        }

        // Mark complete word and bump frequency
        cur.isEnd = true;
        cur.frequency++;
    }

    /*
     * Return ranked autocomplete suggestions for a prefix.
     *
     * Parameters:
     * - prefix: prefix to autocomplete
     * - limit: maximum number of results to return
     *
     * Returns:
     * - List of Map.Entry(word, frequency), best-first
     * - Empty list if prefix is invalid or not found
     *
     * Steps:
     * 1) Walk from root to the node corresponding to the prefix.
     * 2) DFS the subtree under that node, collecting (word, frequency) for each terminal node.
     * 3) Sort collected results by:
     *    - frequency descending
     *    - word ascending for ties
     * 4) Return the first `limit` results (via subList).
     *
     * Note:
     * - This method currently does not guard against non-positive `limit`.
     *   If limit <= 0, Math.min(limit, out.size()) can produce a negative end index,
     *   which would cause an IndexOutOfBoundsException when calling subList().
     *   (No code changes requested—this is just an explanatory note.)
     */
    public List<Map.Entry<String,Integer>> autocomplete(String prefix, int limit) {
        // Walk to prefix node
        Node start = root;

        // Normalize prefix to lowercase and traverse it
        for (char ch : prefix.toLowerCase().toCharArray()) {
            int i = idx(ch);

            // Invalid character or missing path => no suggestions
            if (i < 0 || start.children[i] == null) return List.of();

            start = start.children[i];
        }

        // Collect all completions under the prefix node
        List<Map.Entry<String,Integer>> out = new ArrayList<>();

        // Seed the buffer with the lowercased prefix so DFS can append suffix letters
        dfs(start, new StringBuilder(prefix.toLowerCase()), out);

        // Sort best-first:
        // - higher frequency first
        // - lexicographically smaller word first for ties
        out.sort((a,b) -> {
            if (!a.getValue().equals(b.getValue()))
                return b.getValue() - a.getValue();
            return a.getKey().compareTo(b.getKey());
        });

        // Return only the top `limit` results (or fewer if not enough matches)
        return out.subList(0, Math.min(limit, out.size()));
    }

    /*
     * DFS traversal collecting all terminal words under a subtree.
     *
     * Parameters:
     * - node: current trie node
     * - buf: mutable buffer containing the current word (prefix + explored suffix)
     * - out: output list of (word, frequency) pairs
     *
     * Behavior:
     * - If node.isEnd is true, emit current buffer as a suggestion with node.frequency.
     * - Visit children in 'a'..'z' order:
     *   append letter -> recurse -> backtrack (delete last char).
     *
     * Ordering note:
     * - DFS emits words in lexicographic order, but final output order is determined
     *   by the sort() step in autocomplete().
     */
    private void dfs(Node node, StringBuilder buf,
                     List<Map.Entry<String,Integer>> out) {
        // Emit a suggestion if this node terminates a complete word
        if (node.isEnd)
            out.add(Map.entry(buf.toString(), node.frequency));

        // Explore children in lexicographic order
        for (int i = 0; i < 26; i++) {
            if (node.children[i] != null) {
                // Append next letter
                buf.append((char)('a' + i));

                // Recurse into child subtree
                dfs(node.children[i], buf, out);

                // Backtrack: remove last appended letter
                buf.deleteCharAt(buf.length() - 1);
            }
        }
    }

    /*
     * Load all non-empty lines from a file and insert them into the trie.
     *
     * Parameters:
     * - t: trie instance to populate
     * - path: filesystem path to read
     *
     * Throws:
     * - IOException if file cannot be opened/read
     *
     * Notes:
     * - Uses try-with-resources to ensure the reader is closed.
     * - Does not trim lines; it only checks !w.isEmpty().
     *   Lines containing whitespace-only strings would be treated as "non-empty"
     *   and passed to insert(), which will reject them due to invalid characters.
     *   (No code changes requested—this is just an explanatory note.)
     */
    private static void loadFile(trie_autocomplete_ranked t, String path)
            throws IOException {
        try (BufferedReader br = new BufferedReader(new FileReader(path))) {
            String w;

            // Read each line; insert if non-empty
            while ((w = br.readLine()) != null)
                if (!w.isEmpty()) t.insert(w);
        }
    }

    /*
     * Program entry point.
     *
     * Current behavior:
     * - Uses hard-coded default paths and prefix:
     *   dict = "..\\data\\words.txt"
     *   freq = "..\\data\\frequency.txt"
     *   prefix = "th"
     * - Loads both files into the trie.
     * - Prints the top 20 ranked autocomplete suggestions for the prefix.
     *
     * Exception policy:
     * - Declares `throws Exception`, so I/O and other errors will propagate and
     *   terminate the program unless caught by the runtime.
     */
    public static void main(String[] args) throws Exception {
        trie_autocomplete_ranked trie = new trie_autocomplete_ranked();

        String dict = "..\\data\\words.txt";
        String freq = "..\\data\\frequency.txt";
        String prefix = "th";

        // Load baseline dictionary words, then additional frequency/usage events
        loadFile(trie, dict);
        loadFile(trie, freq);

        // Print ranked suggestions as "word\t(freq=N)"
        for (var e : trie.autocomplete(prefix, 20))
            System.out.println(e.getKey() + "\t(freq=" + e.getValue() + ")");
    }
}
